tests/unit/s2n_key_material_test.c

```diff
@@ -0,0 +1,208 @@
+        s2n_stuffer_write_uint8(&test_data_stuffer, i);
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT helper_validate_key_material(struct s2n_key_material *key_material, struct s2n_blob *test_data_blob,
+        uint8_t mac_size, uint8_t key_size, uint8_t iv_size)
+{
+    /* confirm that the data is copied to key_material */
+    EXPECT_EQUAL(memcmp(test_data_blob->data, key_material->data, test_data_blob->size), 0);
+
+    uint8_t *test_ptr = test_data_blob->data;
+    struct s2n_blob key_material_read = { 0 };
+
+    /* client MAC */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_CLIENT_MAC, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, mac_size), 0);
+    test_ptr += mac_size;
+    /* server MAC */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_SERVER_MAC, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, mac_size), 0);
+    test_ptr += mac_size;
+
+    /* client KEY */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_CLIENT_KEY, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, key_size), 0);
+    test_ptr += key_size;
+    /* server KEY */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_SERVER_KEY, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, key_size), 0);
+    test_ptr += key_size;
+
+    /* client IV */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_CLIENT_IV, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, iv_size), 0);
+    test_ptr += iv_size;
+
+    /* server IV */
+    EXPECT_OK(s2n_key_material_read(key_material, S2N_KEY_MATERIAL_SERVER_IV, &key_material_read));
+    EXPECT_EQUAL(memcmp(test_ptr, key_material_read.data, iv_size), 0);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* prepare test data */
+    uint8_t test_data[S2N_MAX_KEY_BLOCK_LEN] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(helper_generate_test_data(&test_data_blob));
+
+    /* s2n_key_material_init ensure correct size initialization */
+    {
+        struct s2n_key_material key_material = { 0 };
+        DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+
+        struct s2n_record_algorithm record_alg = { 0 };
+        struct s2n_cipher *cipher = NULL;
+
+        /* test stream ciphers */
+        struct s2n_record_algorithm record_alg_stream[] = {
+            s2n_record_alg_rc4_sha
+        };
+        for (int i = 0; i < s2n_array_len(record_alg_stream); i++) {
+            record_alg = record_alg_stream[i];
+            cipher = (struct s2n_cipher *) record_alg.cipher;
+            conn->secure->cipher_suite->record_alg = &record_alg;
```

Comment:
`conn->secure->cipher_suite` would be s2n_null_cipher_suite since you haven't performed a handshake. By doing this, you're modifying the record alg of s2n_null_cipher_suite for this entire process, which is extremely dangerous.

Reply:
created a temp s2n_cipher_suite and modifying that